test/extended/util/disruption/controlplane/termination.go

```diff
@@ -0,0 +1,120 @@
+
+	"k8s.io/client-go/tools/watch"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+)
+
+// GracefulTerminationTest watches events of non-graceful terminations of kube-apiserver.
+type GracefulTerminationTest struct{}
+
+func (GracefulTerminationTest) Name() string { return "kube-apiserver-terminates-gracefully" }
+func (GracefulTerminationTest) DisplayName() string {
+	return "Kubernetes API server terminates gracefully"
+}
+func (GracefulTerminationTest) Setup(f *framework.Framework)    {}
+func (GracefulTerminationTest) Teardown(f *framework.Framework) {}
+func (t *GracefulTerminationTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		defer cancel()
+		<-done
+	}()
+
+	m := EventMonitor{
+		Namespace: "openshift-kube-apiserver",
+		Filter: func(ev *corev1.Event) bool {
+			return ev.Reason == "NonGracefulTermination"
+		},
+		Handler: func(f *framework.Framework, ev *corev1.Event) {
+			framework.Failf("kube-apiserver reports a non-graceful termination: %#v", ev)
+		},
+	}
+	m.Run(ctx, f)
+}
+
+// LateConnectionsTest watches events late connections to the kube-apiserver during termination. These usually
+// mean that the load balancer do not configure their health probes correctly, by calling /readyz of the API
+// server (not /healthz !)
+type LateConnectionsTest struct{}
+
+func (LateConnectionsTest) Name() string { return "kube-apiserver-no-late-connections" }
+func (LateConnectionsTest) DisplayName() string {
+	return "LBs stop connecting to kube-apiserver during termination in time"
+}
+func (LateConnectionsTest) Setup(f *framework.Framework)    {}
+func (LateConnectionsTest) Teardown(f *framework.Framework) {}
+func (t *LateConnectionsTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		defer cancel()
+		<-done
+	}()
+
+	m := EventMonitor{
+		Namespace: "openshift-kube-apiserver",
+		Filter: func(ev *corev1.Event) bool {
+			return ev.Reason == "LateConnections"
+		},
+		Handler: func(f *framework.Framework, ev *corev1.Event) {
+			framework.Failf("kube-apiserver got connections late during termination, probably due to broken LB configuration: %#v", ev)
+		},
+	}
+	m.Run(ctx, f)
+}
+
+// EventMonitor watches events in the given namespace, calls the filter (optionally, can be nil)
+// and calls Handler for those that the filter returned true for (or all if filter is nil).
+type EventMonitor struct {
+	Namespace string
+	Filter    func(event *corev1.Event) bool
+	Handler   func(f *framework.Framework, event *corev1.Event)
+}
+
+func (m *EventMonitor) Run(ctx context.Context, f *framework.Framework) {
+	config, err := framework.LoadConfig()
+	framework.ExpectNoError(err)
+
+	c, err := kubernetes.NewForConfig(config)
+	framework.ExpectNoError(err)
+
+	lw := &cache.ListWatch{
```

Comment:
Is there really an advantage over this with just running "get all events, filter to see if they match" at the end of the test (which is both simpler and can be easily copied into a `[Late]` test)?  Upgrades take less than 2hr normally, so events aren't going to TTL.  If we lose the DB, we have bigger problems.  The events won't be in namespaces that get deleted.

Reply:
Moved into `[Late]` test.